### PR TITLE
[Research] Adding the ViGIL experiments tutorials

### LINKED ISSUE #104: Not all loggers output are captured in the GridWorkers
When using a `GridWorker`, only the output of the logger of the current `Worker` (e.g. `OfflineTrainer` or `Tester`) is captured. The loggers of the `Model` & `Problem` are not able to write to the console or to the `.log` file. 
This is fairly annoying as these loggers have some important information (e.g. whether the `Sampler` was properly instantiated, when is saved the `Model` etc).

Not sure yet how to fix this. I think this is related to the `logging.config` and also how the output of a child process is captured in the `GridWorker`:

https://github.com/IBM/mi-prometheus/blob/3d4f868320c8914f37543298c634ab6ba6a1425d/miprometheus/grid_workers/grid_trainer_cpu.py#L316-L317

We should probably change `stdout=os.devnull` and do something instead like

`with open(`**worker.log**`, 'w') as devnull: `
              `result = subprocess.run(command_str.split(" "), stdout=devnull) `

### PR DESCRIPTION
**Pull request by vmarois**
In this PR, I am adding a _Reproducible Research_ [section](https://mi-prometheus.readthedocs.io/en/research-vigil/research/2_vigil.html) in the documentation. This section only contains for now a note on how to reproduce the experiments described in the ViGIL paper, which can be taken as an example for the future notes in this section.

**Main points:**
- Added the S-MAC model to `miprometheus.models`. It reuses as many units as possible from the MAC model implementation, such that it only implements the units which are different in terms of equations. Its documentation reflects the published paper (bibtex is indicated).

- Added all the grid config files to run the experiments:
        - The initial training on CLEVR & CoGenT-A,
        - The finetuning on CoGenT-B of the CLEVR- & CoGenT-A-trained models,
        - The finetuning on CoGenT-A of the CLEVR-trained models,

Default configuration files are available for CLEVR / CogenT / MAC on CLEVR / MAC on CoGenT / S_MAC on CLEVR / S-MAC on CoGenT. The grid config files reuse these. The grid config files are pretty long and complex because the configuration for each experiment is different, but everything should be commented.
- All tests experiment are indicated in the grid config files with the ``multi_tests`` key (Was able to do so after the merge of #98 and #100). Adding support for multi-tests in the `Tester` was long to implement, but is very useful here as we can run "cross-tests" (e.g. train on CoGenT-A and test on CoGenT-B) with the same command as running "regular tests" (e.g. train on CoGenT-A and test on CoGenT-A). 
- To allow the proper test of a CLEVR-trained model on CoGenT samples, we need to make sure that the dicts `{'words': index}` &  `{'answer': index}` are the same for both CoGenT & CLEVR, but also that the random embedding weights are the same. I formalized that under the form of an additional param for the `CLEVR` class (`embedding_source`, doc has been updated, and the `CLEVR` class throws a warning).

Overall, the pipeline is to run `mip-grid-trainer-gpu`, `mip-grid-tester-gpu` and `mip-grid-analyzer` 3 times. 

This documentation section should be self-consistent, in that all config files (and indices files for the `Sampler`) are linked (I provide ours for transparency). All commands are indicated.

**note**:
- I am not using Exponential Moving Average as this is not implemented yet (was done on the internal repo). I am also not sure that this is making a major difference in performance. We may want to implement this in the future.

I have tested the overall and it should be working. I cannot 100% guarantee that I squashed every possible bug here, but most should be fixed :slightly_smiling_face: 

:warning: A few things do need to be fixed to ensure that this can reproduced anywhere:
- Ensure that when indicating `default_configs`, the path to these files can be properly constructed from the path of the specified initial config file (with `--c`). I gave a quick shot at this, and it may not be as straightforward as initially thought: we need to handle the case when we specify several config files (separated by commas), and also in the `GridWorkers` when `Named Temporary Files` are created and then superseded by the `config params`. Cf #16,
- Automate the downloads of the dataset files (cf #101 ),
- Resolve #104, i.e. ensure that all loggers (`Problem`, `Model`, `SamplerFactory`...) can properly log to the console and the log file (although this is not critical),
- Ensure that when starting a grid experiment, each single experiment is moved to one GPU (link #37). **I have set the `sleep` time to 60s in this PR.**

### CODE REVIEW

File: configs/mac/mac_clevr.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Load the CLEVR config first.
+default_configs:
+    configs/mac/default_clevr.yaml
+
+# Add the model parameters:
+model:
```
Comment (tkornut) [inline configs/mac/mac_clevr.yaml]: Maybe we could move model to default_mac.yaml as well?

File: configs/mac/mac_smac_cogent_a_finetuning.yaml [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    seed_torch: 0
+    seed_numpy: 0
+
+  validation:
+    problem:
+      settings:
+        set: 'trainA'  # use CoGenT-A as validation set (-> 10% of the true training set) as we will test on CoGenT.
+    sampler:
+      name: 'SubsetRandomSampler'
+      indices: '~/data/CLEVR_CoGenT_v1.0/vigil_cogent_val_set_indices.txt'
```
Comment (tkornut) [inline configs/mac/mac_smac_cogent_a_finetuning.yaml]: Is this full validation set?

File: configs/mac/mac_clevr.yaml (hunk 1 above)
Comment (vmarois) [reply configs/mac/mac_clevr.yaml]: Given that these experiments are on 2 models (MAC & S-MAC), not sure it's very useful
